### PR TITLE
Add NLP example: SST-2

### PR DESCRIPTION
**Pull request by bastings**
This adds binary Stanford Sentiment classification (SST-2) as used in the GLUE benchmark.
The model is a LSTM classifier and serves as a simple example of how to use Flax for NLP.

### CODE REVIEW

File: examples/sst2/README.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+{
+  'sentence': 'this cross-cultural soap opera is painfully formulaic and stilted .',
+  'label': 0
+}
+```
+
+In this example we will build and train a **text classifier** on this data set. 
+
+Each sentence is mapped to a sequence of integers using a vocabulary.
+The vocabulary is just a dictionary that maps each word that occurs in SST-2 to a unique ID.  That looks like this for the above sentence:
```
Comment (marcvanzee) [inline examples/sst2/README.md]:
```suggestion
The vocabulary is just a dictionary that maps each word that occurs in SST-2 to a unique ID. It looks as follows for the above sentence:
```

File: examples/sst2/README.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+```python
+{
+  'sentence':  [2, 4, 28, 10199, 3817, 4527, 13, 7380, 11530, 7, 12366, 5, 3],
+  'label': 0
+}
+```
+Note that a `2` was added at the beginning to mark the start, and a `3` to mark the end of the sequence.
+
+## Model
+Our model consists of word embeddings, an LSTM encoder, and an MLP classifier.
```
Comment (marcvanzee) [inline examples/sst2/README.md]: This drawing is great! :)

File: examples/sst2/README.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+## Historic note on SST-2
+This dataset consisted of annotated *trees* originally, with sentiment labels (very negative to very positive) at every node of the tree.  There are only 6920 training trees with non-neutral root labels, and to make more use of the  dataset the yields of sub-trees were also included in the training set. (Exactly how that was done seems to be unknown at this point.) So that means that the training data contains a lot of overlapping phrases!
+
+## Instructions
+
+1. Download and unzip the pre-trained word embeddings:
+
+```sh
+wget http://nlp.stanford.edu/data/glove.840B.300d.zip
```
Comment (marcvanzee) [inline examples/sst2/README.md]: Maybe say something about the file size? It seems to be quite large.

File: examples/sst2/README.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+unzip glove.840B.300d.zip
+```
+
+2. Train the model as follows. The SST-2 data is downloaded from TFDS and cached
+   on the first run.
+
+```sh
+python3 train.py --glove-path glove.840B.300d.txt
```
Comment (marcvanzee) [inline examples/sst2/README.md]: On which platform did you run it? Alexey spent quite some time recently ensuring all examples have the information in the README.md, except for very simple examples. E.g.:
- https://github.com/google/flax/tree/prerelease/examples/lm1b
- https://github.com/google/flax/tree/prerelease/examples/imagenet

Or is your example so simple that it can be run on a CPU?

File: examples/sst2/input_pipeline.py [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+import os
+
+from absl import logging
+
+import numpy as np
+import tensorflow.compat.v2 as tf
+import tensorflow_datasets as tfds
+
+
+# pylint: disable=no-member
```
Comment (marcvanzee) [inline examples/sst2/input_pipeline.py]: Why do we need this?

File: examples/sst2/input_pipeline.py [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+  for ds in datasets:  # Data set already tokenized here.
+    for example in tfds.as_numpy(ds):
+      tokens = whitespace_tokenize(example['sentence'].strip())
+      counter.update(tokens)
+  sst_tokens = set(counter.keys())
+  logging.info('Number of unique tokens: %d', len(sst_tokens))
+  return sst_tokens
+
+
+def save_vocab(vocab, path):
```
Comment (marcvanzee) [inline examples/sst2/input_pipeline.py]: Currently half of your code is typed, and the other half isn't. Could you please make it typed everywhere?

File: examples/sst2/input_pipeline.py [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+  with open(path, mode='wb') as f:
+    for token in vocab:
+      f.write(token + b'\n')
+  logging.info('Saved vocab to: %s', path)
+
+
+def load_vocab(path):
+  """Loads a vocabulary from disk."""
+  vocab = collections.OrderedDict()
+  with open(path, mode='rb') as f:
```
Comment (marcvanzee) [inline examples/sst2/input_pipeline.py]: I think 'r' is the default so you can omit.

```suggestion
  with open(path, mode='b') as f:
```

But feel free to keep it if you think it is clearer...

File: examples/sst2/input_pipeline.py [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+  np.savez(path, embed=embed)
+  logging.info('Saved embed to: %s', path)
+
+
+def load_embed(path):
+  """Load word embeddings from disk."""
+  return np.load(path)['embed']
+
+
+# pylint: disable=dangerous-default-value
```
Comment (marcvanzee) [inline examples/sst2/input_pipeline.py]: I guess this is safe since "specials" is not modified in the function. However, it is a bit dangerous practice: https://stackoverflow.com/questions/26320899/why-is-the-empty-dictionary-a-dangerous-default-value-in-python

Maybe add a note that this is safe since we don't modify "specials"?

File: examples/sst2/input_pipeline.py [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+      os.mkdir(cache_dir)
+    save_vocab(vocab, os.path.join(cache_dir, 'vocab.txt'))
+    save_embed(embed, os.path.join(cache_dir, 'embed.npz'))
+
+  return vocab, embed
+
+
+def whitespace_tokenize(s: Text):
+  """Splits ab input into tokens by whitspace."""
+  return s.split()
```
Comment (marcvanzee) [inline examples/sst2/input_pipeline.py]:
```suggestion:
def whitespace_tokenize(input: Text):
  """Splits input by into tokens by whitespace."""
  return input.split()
```

File: examples/sst2/input_pipeline.py (hunk 7 above)
Comment (bastings) [reply examples/sst2/input_pipeline.py]: Maybe I'll keep 'rb' if you're OK with it.

File: examples/sst2/input_pipeline.py (hunk 6 above)
Comment (bastings) [reply examples/sst2/input_pipeline.py]: done

File: examples/sst2/input_pipeline.py (hunk 5 above)
Comment (bastings) [reply examples/sst2/input_pipeline.py]: TF has a strange way of defining inheritance and pylint doesn't get it. (Some linters don't check this deep though.)

File: examples/sst2/README.md (hunk 4 above)
Comment (bastings) [reply examples/sst2/README.md]: It can run on CPU too, but I'll see if  I can add more info.

File: examples/sst2/input_pipeline.py (hunk 8 above)
Comment (bastings) [reply examples/sst2/input_pipeline.py]: Fixed

File: examples/sst2/input_pipeline.py (hunk 9 above)
Comment (bastings) [reply examples/sst2/input_pipeline.py]: fixed
